### PR TITLE
Add code coverage

### PR DESCRIPTION
**Pull request by amelie-renard-sonarsource**

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -12,13 +12,29 @@ jobs:
...
     displayName: Download and install build wrapper
   - pwsh: |
       $path = vswhere -latest -requires Microsoft.Component.MSBuild -find MSBuild\**\Bin\MSBuild.exe | select-object -first 1
       & "$HOME\.sonar\build-wrapper-win-x86\build-wrapper-win-x86-64.exe" --out-dir build_wrapper_output_directory $path sonar_scanner_example.vcxproj /t:rebuild /nodeReuse:false
+      # Execute the tests to generate the coverage statistics
+      & $path test1\test1.vcxproj
+      & $path test2\test2.vcxproj
     workingDirectory: .
-    displayName: Build in build-wrapper
+    displayName: Build in build-wrapper and tests
```
Comment (arseniy-sonar) [inline azure-pipelines.yml:28]:
```suggestion
    displayName: Build in build-wrapper and run tests
```

File: azure-pipelines.yml [hunk 2]
```diff
@@ -12,13 +12,29 @@ jobs:
...
+    inputs:
+      testSelector: 'testAssemblies'
+      testAssemblyVer2: |
+        test1\*\test1.dll
+        test2\*\test2.dll
+      searchFolder: '$(System.DefaultWorkingDirectory)'
+      resultsFolder: '$(System.DefaultWorkingDirectory)/CoverageResults'
+      otherConsoleOptions: '/collect:"Code Coverage;Format=Xml"' # Necessary to have an XML coverage report
+      codeCoverageEnabled: true
+      testRunTitle: 'Run tests'
```
Comment (arseniy-sonar) [inline azure-pipelines.yml:39]: Does this run the tests? What do lines 25 and 26 do then?

**Review by arseniy-sonar — APPROVED** [review]

**Comment by arseniy-sonar** [comment]
LGTM
The coverage collection looks quite specific to Azure. I think we need another example that shows how to do it in the shell (the vscoverage-otherci ?)

File: azure-pipelines.yml (hunk 2 above)
Comment (amelie-renard-sonarsource) [reply azure-pipelines.yml:39]: Yes, this runs the tests
l.25 and l.26 build the two test projects

**Comment by amelie-renard-sonarsource** [comment]
I agree, the "VSTest" is really specific to Azure, the hard part here was to find how to convert the ".coverage" format to a XML format. 
I think additional example is necessary indeed

File: src/lib.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#include "lib.h"
+
+int myfun(bool b, int (*callback)(int)) {
+  int num = -1;
+
+  if (b) {
+    num = callback(num);
+  } else {
+    return callback(callback(num));
+  }
```
Comment (arseniy-sonar) [inline src/lib.cpp:10]:
```suggestion
  if (b) {
    num = callback(callback(num));
  } else {
    return num;
  }
```
To trigger the nullptr dereference on line 18, need to remove a call to `callback` at least from one branch.

**Review by arseniy-sonar — APPROVED** [review]

**Pull request merged**